gluonnlp/vocab.py

```diff
@@ -456,3 +471,370 @@ def from_json(json_str):
+
+        """
+
+        return [self.idx_to_subwordidxs_list[i] for i in indices]
+
+    def indices_to_subwordindices_mask(self, indices):
+        """Return array of subwordindices for indices.
+
+        A padded numpy array and a mask is returned. The mask is used as
+        indices map to varying length subwords.
+
+        Parameters
+        ----------
+        indices : list of int, numpy array or mxnet NDArray
+            Token indices that should be mapped to subword indices.
+
+        Returns
+        -------
+        Array of subword indices.
+
+        """
+        if isinstance(indices, nd.NDArray):
+            indices = indices.asnumpy().astype(np.int)
+        else:
+            indices = np.array(indices, dtype=np.int)
+        subwords = self.idx_to_subwordidxs[indices]
+        mask = np.zeros_like(subwords)
+        mask += subwords != -1
+        lengths = np.argmax(subwords == -1, axis=1)
+
+        new_length = max(np.max(lengths), 1)  # Return at least 1
+        subwords = subwords[:, :new_length]
+        mask = mask[:, :new_length]
+
+        return subwords, mask
+
+    def words_to_subwordindices(self, words):
+        """Return list of lists of subwordindices for words.
+
+        Parameters
+        ----------
+        indices : iterable of str
+            Words that should be mapped to subword indices.
+
+        Returns
+        -------
+        List of variable length lists of subword indices.
+
+        """
+
+        subwordindices = list(self.subword_function(words))
+        return subwordindices
+
+    def subwordindices_to_subwords(self, subwordindices):
+        """Return list of subwords associated with subword indices.
+
+        This may raise RuntimeError if the subword function is not invertible.
+
+        Parameters
+        ----------
+        subwordindices : iterable of int
+            Subword indices to look up.
+
+        Returns
+        -------
+        Iterable of str.
+
+        """
+
+        subwordindices = subwordindices
+        return self.subword_function.indices_to_subwords(subwordindices)
+
+    def subwords_to_subwordindices(self, subwords):
+        """Return list of subwordindices associated with subwords.
+
+        Parameters
+        ----------
+        subwords : iterable of str
+            Subwords to replace by indices.
+
+        Returns
+        -------
+        Iterable of int.
+
+        """
+        return self.subword_function.subwords_to_subwordindices(subwords)
+
+    def __len__(self):
+        return len(self.subword_function)
+
+    def to_json(self):
+        """Serialize subword vocab object to json string."""
+        dict_ = {}
+        try:
+            dict_['subwordidx_to_subword'] = \
+                self.subword_function.indices_to_subwords(
+                    list(range(len(self.subword_function))))
+        except RuntimeError:
+            # Not all subword functions are invertible
+            pass
```

Comment:
silently failing might cause weird bugs here.

Also, I don't understand the logic here. why would subword_function have length? serializing a function into json is very tricky...

Reply:
If we limit the subword functions to the available options, then it becomes easier. we just need to save the type string of the subword function, along with all the constructor arguments.